### PR TITLE
added input to data source

### PR DESCRIPTION
**Pull request by datanel**
```json
{
  "id": "tram_lyon",
  "name": "Tram feed for Lyon",
  "data_format": "gtfs",
  "input": [
    {
      "type": "url",
      "url": "http://stif.com/od.zip"
    }
  ]
}
```

- [x] Update Doc
- [x] Refacto tests

### CODE REVIEW

File: tartare/core/models.py [hunk 1]
```diff
@@ -250,10 +252,33 @@ def get_contributor(cls, contributor_id):
...
+class Input(object):
+    def __init__(self, _type=None, url=None, v=None):
+        self._type = _type
+
+        if url:
+            self.url = url
+
+        if v:
+            self.v = v
+
```
Comment (azime) [inline tartare/core/models.py]: You can not use (key, value) ?
```json
{'input': {'key' : 'url', 'value': 'stif.com/od.zip'}}
```
```json
{'input': {'key' : 'existing_version', '-2'}}
```

File: tartare/core/models.py [hunk 2]
```diff
@@ -176,13 +176,15 @@ def remove_none(self, data):
 
 
 class DataSource(object):
-    def __init__(self, id=None, name=None, data_format="gtfs"):
+    def __init__(self, id=None, name=None, data_format="gtfs", data_prefix=None, input={}):
```
Comment (xlqian) [inline tartare/core/models.py]: hmm, it'd be better not to use object as the default value of the arguments, which may lead to some unexpected behaviors :)

```python
def function(data=[]):
     data.append(1)
     return data

function()
>>>[1]
function()
>>>[1, 1]
function()
>>>[1, 1, 1]
```

File: tests/integration/mongo/conftest.py [hunk 3]
```diff
@@ -92,7 +92,8 @@ def contributor(app):
 def data_source(app, contributor):
     data_source = app.post('/contributors/{}/data_sources'.format(contributor.get('id')),
                            headers={'Content-Type': 'application/json'},
-                           data='{"name": "bobette", "data_format": "gtfs"}')
+                           data='{"name": "bobette", "data_format": "gtfs", "data_prefix": "BOB", "input": {"type": "url", "url": "http://bob.com"}}')
+    print(data_source)
```
Comment (azime) [inline tests/integration/mongo/conftest.py]: remove ?

File: tartare/core/models.py [hunk 4]
```diff
@@ -250,10 +252,33 @@ def get_contributor(cls, contributor_id):
...
+    @post_load
+    def build_input(self, data):
+        return Input(**data)
+
+
 class MongoDataSourceSchema(Schema):
     id = fields.String(required=True)
     name = fields.String(required=True)
     data_format = fields.String(required=False)
+    data_prefix = fields.String(required=False, allow_none=True)
```
Comment (azime) [inline tartare/core/models.py]: required=True ?

**Comment by kadhikari** [comment]
Can you add the last test added in the pr: https://github.com/CanalTP/tartare/pull/87
"test_post_ds_one_data_source_without_input" please?

File: tartare/core/models.py [hunk 5]
```diff
@@ -177,13 +177,15 @@ def remove_none(self, data):
...
 class DataSource(object):
-    def __init__(self, id=None, name=None, data_format="gtfs"):
+    def __init__(self, id=None, name=None, data_format="gtfs", data_prefix=None, input=[]):
         if not id:
             self.id = str(uuid.uuid4())
         else:
             self.id = id
         self.name = name
         self.data_format = data_format
+        self.data_prefix = data_prefix
```
Comment (nberard) [inline tartare/core/models.py]: still used?

File: tests/integration/mongo/data_sources_api_test.py [hunk 6]
```diff
@@ -70,7 +74,9 @@ def test_post_ds_one_data_source_with_id(app):
     post_contrib = {"id": "id_test", "name":"name_test", "data_prefix":"AAA"}
     raw = post(app, '/contributors', json.dumps(post_contrib))
     assert raw.status_code == 201, print(to_json(raw))
-    post_ds = {"id": "data_source_id", "name":"data_source_name"}
+    post_ds = {"id": "data_source_id", "name": "data_source_name",
+               "input": [{"key": "type", "value": "url"},
+                                               {"key": "url", "value": "http://stif.com/od.zip"}]}
```
Comment (pbougue) [inline tests/integration/mongo/data_sources_api_test.py]: minor: indent

File: tests/integration/mongo/data_sources_api_test.py [hunk 7]
```diff
@@ -331,3 +402,22 @@ def test_patch_contrib_one_data_source_name_of_two_and_add_one(app):
...
+    post_ds = {"name": "data_source_name"}
+    raw = post(app, '/contributors/id_test/data_sources', json.dumps(post_ds))
+    r = to_json(raw)
+    assert 'error' in r
+    assert raw.status_code == 400, print(to_json(raw))
+
+    raw = app.get('/contributors/id_test/data_sources')
+    r = to_json(raw)
+    assert raw.status_code == 200, print(r)
+    assert len(r["data_sources"]) == 0
```
Comment (pbougue) [inline tests/integration/mongo/data_sources_api_test.py]: minor: missing newline

File: tartare/core/models.py (hunk 5 above)
Comment (pbougue) [inline tartare/core/models.py]: I vote remove it

**Comment by datanel** [comment]
@pbougue your tests ;)

File: tartare/core/models.py [hunk 8]
```diff
@@ -177,13 +177,14 @@ def remove_none(self, data):
 
 
 class DataSource(object):
-    def __init__(self, id=None, name=None, data_format="gtfs"):
+    def __init__(self, id=None, name=None, data_format="gtfs", data_prefix=None, input=[]):
```
Comment (azime) [inline tartare/core/models.py]: Unused argument "data_prefix"

File: tests/integration/mongo/data_sources_api_test.py [hunk 9]
```diff
@@ -239,19 +280,26 @@ def test_patch_ds_data_source_name_only(app):
...
+    post_data["data_sources"].append({"name":"data_source_name", "data_format":"Neptune",
+                                      "input": [{"key": "type", "value": "url"},
+                                                {"key": "url", "value": "http://stif.com/od.zip"}]})
     raw = post(app, '/contributors', json.dumps(post_data))
     r = to_json(raw)
     assert raw.status_code == 201, print(r)
     new_data_source = {}
     new_data_source["id"] = r["contributors"][0]["data_sources"][0]["id"]
     new_data_source["name"] = "name_modified"
+    new_data_source["data_prefix"] = "LOL"
```
Comment (azime) [inline tests/integration/mongo/data_sources_api_test.py]: to remove, no ?

File: tartare/core/models.py [hunk 10]
```diff
@@ -251,10 +252,26 @@ def get_contributor(cls, contributor_id):
...
 
+class Input(object):
+    def __init__(self, key=None, value=None):
+        self.key = key
+        self.value = value
+
+
+class MongoInputSchema(Schema):
+    key = fields.String(required=True)
+    value = fields.String(required=True)
```
Comment (kinnou02) [inline tartare/core/models.py]: seriously?

Comment (datanel) [inline tartare/core/models.py]: absolutely

File: tests/integration/mongo/contributors_api_test.py [hunk 11]
```diff
@@ -175,3 +176,241 @@ def test_update_contributor_id_impossible(app):
...
+                    "url": "http://stif.com/od.zip"
+                }
+            }
+        ]
+    }
+    raw = post(app, '/contributors', json.dumps(post_data))
+    assert raw.status_code == 201, print(to_json(raw))
+    raw = app.get('/contributors/id_test/')
+    r = to_json(raw)
+    print(r)
```
Comment (azime) [inline tests/integration/mongo/contributors_api_test.py]: to remove ?

**Pull request merged**